### PR TITLE
Markusahlstrand/sc 4681/make account linking work like in auth0

### PR DESCRIPTION
**Pull request by markusahlstrand**
This adds the link step by filtering the users based on the linked_to field and querying for linked users.

### CODE REVIEW

File: integration-test/management-api/users.spec.ts [hunk 1]
```diff
@@ -52,6 +52,7 @@ describe("users", () => {
       method: "POST",
       body: JSON.stringify({
         username: "test@example.com",
+        connection: "email",
```
Comment (markusahlstrand) [inline integration-test/management-api/users.spec.ts:55]: The connections seems to be required in auth0. Guess we need that as the userId depends on it?

File: integration-test/management-api/users.spec.ts [hunk 2]
```diff
@@ -95,5 +97,112 @@ describe("users", () => {
...
+      const listUsersResponse = await worker.fetch("/api/v2/users", {
+        headers: {
+          authorization: `Bearer ${token}`,
+          "tenant-id": "test",
+        },
+      });
+
+      expect(listUsersResponse.status).toBe(200);
+
+      const usersList = await listUsersResponse.json();
```
Comment (markusahlstrand) [inline integration-test/management-api/users.spec.ts:173]: This is cheating slightly.. The identities of the listed users aren't set correctly when using the in-memory adapter.

File: src/adapters/in-memory/users/index.ts [hunk 3]
```diff
@@ -3,7 +3,7 @@ import { UserDataAdapter } from "../../interfaces/Users";
 import { createUser } from "./create";
 import { get } from "./get";
 import { getByEmail } from "./getByEmail";
-import { listUsers } from "./listUsers";
+import { list } from "./list";
```
Comment (markusahlstrand) [inline src/adapters/in-memory/users/index.ts:6]: changed for consitency

File: src/adapters/in-memory/users/list.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import { ListParams } from "../../interfaces/ListParams";
+import { filterItems } from "../utils/filter";
+
+export function list(users: SqlUser[]) {
+  return async (
+    tenantId,
+    { page, per_page, include_totals, q }: ListParams,
+  ): Promise<ListUsersResponse> => {
+    const tenantUsers = users.filter((user) => user.tenant_id === tenantId);
+    const filteredUsers = filterItems(tenantUsers, q);
```
Comment (markusahlstrand) [inline src/adapters/in-memory/users/list.ts:12]: Apply a happy-path lucene filter

File: src/adapters/in-memory/utils/filter.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
+export function filterItems<T extends object>(items: T[], query?: string): T[] {
```
Comment (markusahlstrand) [inline src/adapters/in-memory/utils/filter.ts:1]: A very simplistic lucene filter. Only handles the simple cases. We need to do this for sql as well...

File: src/routes/management-api/users.ts [hunk 6]
```diff
@@ -51,28 +51,35 @@ export class UsersMgmtController extends Controller {
...
+      q: query.join(" "),
     });
 
     const users: UserResponse[] = result.users.map((user) => {
-      const identities = [];
-
       return {
         ...user,
         user_id: user.id,
-        logins_count: 0,
```
Comment (markusahlstrand) [inline src/routes/management-api/users.ts:69]: These are part of the user now

File: src/routes/management-api/users.ts [hunk 7]
```diff
@@ -96,24 +103,36 @@ export class UsersMgmtController extends Controller {
...
     }
 
+    const linkedusers = await env.data.users.list(tenant_id, {
+      page: 0,
+      per_page: 10,
+      include_totals: false,
+      q: `linked_to:${user_id}`,
+    });
+
+    const identities = [user, ...linkedusers.users].map((u) => ({
```
Comment (markusahlstrand) [inline src/routes/management-api/users.ts:124]: Fetch and linked users

File: src/routes/management-api/users.ts [hunk 8]
```diff
@@ -96,24 +103,36 @@ export class UsersMgmtController extends Controller {
...
 
     if (!user) {
       throw new HTTPException(404);
     }
 
+    const linkedusers = await env.data.users.list(tenant_id, {
+      page: 0,
+      per_page: 10,
+      include_totals: false,
+      q: `linked_to:${user_id}`,
```
Comment (markusahlstrand) [inline src/routes/management-api/users.ts:121]: Use a lucene query to fetch linked users.

File: src/routes/management-api/users.ts [hunk 9]
```diff
@@ -163,7 +182,7 @@ export class UsersMgmtController extends Controller {
       connection: "email",
       email_verified: false,
       last_ip: "",
-      login_count: 0,
+      logins_count: 0,
```
Comment (markusahlstrand) [inline src/routes/management-api/users.ts:185]: Right.. we hade a typo here. Need to make a migration for this

File: src/routes/management-api/users.ts [hunk 10]
```diff
@@ -211,70 +234,35 @@ export class UsersMgmtController extends Controller {
...
+    const { env } = request.ctx;
+
+    const user = await env.data.users.get(tenantId, body.link_with);
+    if (!user) {
+      throw new HTTPException(400, {
+        message: "Linking to an inexistent identity is not allowed.",
+      });
+    }
+
+    await env.data.users.update(tenantId, userId, {
```
Comment (markusahlstrand) [inline src/routes/management-api/users.ts:247]: This is doing the actual linking

File: src/types/auth0/Identity.ts [hunk 11]
```diff
@@ -6,7 +6,7 @@ export interface Identity {
   access_token?: string;
   access_token_secret?: string;
   refresh_token?: string;
-  profileData: {
+  profileData?: {
```
Comment (markusahlstrand) [inline src/types/auth0/Identity.ts:9]: This seems to be optional, right?

File: test/adapters/in-memory/utils/filter.spec.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
+const {
+  filterItems,
+} = require("../../../../src/adapters/in-memory/utils/filter");
+
+describe("filter", () => {
```
Comment (markusahlstrand) [inline test/adapters/in-memory/utils/filter.spec.ts:5]: Validating that the in-memory filter works...

File: integration-test/management-api/users.spec.ts (hunk 1 above)
Comment (danielgent) [reply integration-test/management-api/users.spec.ts:55]: Confirmed I see this!  But `provider` is not required

File: integration-test/management-api/users.spec.ts [hunk 13]
```diff
@@ -95,5 +97,112 @@ describe("users", () => {
     const body = await usersResponse.json();
     expect(body.length).toBe(1);
     expect(body[0].user_id).toBe(newUser.user_id);
```
Comment (danielgent) [inline integration-test/management-api/users.spec.ts:99]: :eyes: This is what I started on https://github.com/sesamyab/auth/pull/259 - the concatenation of the provider + id

Happy to sync on this first. Was just getting my thoughts down in code :smile:

File: src/adapters/in-memory/users/index.ts (hunk 3 above)
Comment (danielgent) [reply src/adapters/in-memory/users/index.ts:6]: But without using `git mv` so GitHub shows the whole file as one diff

it doesn't take long to write `git checkout -b feat/do-this-cleanup-on-a-newpr`

File: src/types/auth0/Identity.ts (hunk 11 above)
Comment (danielgent) [reply src/types/auth0/Identity.ts:9]: I see this. Only present on any linked profile identities 

image

File: src/routes/management-api/users.ts (hunk 9 above)
Comment (danielgent) [reply src/routes/management-api/users.ts:185]: Cool I can do this now :+1: 

I can do this change on just that PR. Better to keep it simple

`login_count` needs to be `logins_count` everywhere

Comment (danielgent) [reply src/routes/management-api/users.ts:185]: Do we need to migrate the values we already have in `login_count`? I don't think we've populated anything yet :smile:

File: test/adapters/in-memory/utils/filter.spec.ts [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        description: "Outdated model",
+      },
+    ]);
+  });
+
+  test("filters items where a property does not exist", () => {
+    const result = filterItems(items, "-_exists_:description");
+    expect(result).toEqual([{ category: "books", status: "inactive" }]);
+  });
+});
```
Comment (danielgent) [inline test/adapters/in-memory/utils/filter.spec.ts:83]: Nice! :fire:

File: integration-test/management-api/users.spec.ts (hunk 13 above)
Comment (markusahlstrand) [reply integration-test/management-api/users.spec.ts:99]: Yes, we need to get the user-id working consistently!

File: src/routes/management-api/users.ts (hunk 9 above)
Comment (markusahlstrand) [reply src/routes/management-api/users.ts:185]: I don't think we actually increate it yet :)

File: integration-test/management-api/users.spec.ts [hunk 15]
```diff
@@ -95,5 +97,112 @@ describe("users", () => {
...
+          authorization: `Bearer ${token}`,
+          "tenant-id": "test",
+        },
+      });
+
+      expect(listUsersResponse.status).toBe(200);
+
+      const usersList = await listUsersResponse.json();
+      expect(usersList.length).toBe(1);
+      expect(usersList[0].user_id).toBe(newUser2.user_id);
```
Comment (danielgent) [inline integration-test/management-api/users.spec.ts:175]: I'm not following what this proves at all :joy: 

You're setting `newUser2` as the _primary_ user? (maybe call it `primaryUser`? Or we come up with some terminology... maybe Auth0 has a name for this)

File: src/adapters/in-memory/users/list.ts (hunk 4 above)
Comment (markusahlstrand) [reply src/adapters/in-memory/users/list.ts:12]: Yeah.. the rename of the file throw it off..

File: src/adapters/in-memory/users/index.ts (hunk 3 above)
Comment (markusahlstrand) [reply src/adapters/in-memory/users/index.ts:6]: Ahh, nice trick!

File: integration-test/management-api/users.spec.ts [hunk 16]
```diff
@@ -95,5 +97,112 @@ describe("users", () => {
...
+          connection: "email",
+          user_id: newUser2.user_id,
+          provider: "email",
+          isSocial: false,
+        },
+        {
+          connection: "email",
+          user_id: newUser1.user_id,
+          provider: "email",
+          isSocial: false,
```
Comment (danielgent) [inline integration-test/management-api/users.spec.ts:203]: here `profileData` needs to be included though!

That's what I see from Auth0. We could even type this in TypeScript I think... have the first element as one type
```
 connection: "email",
user_id: newUser2.user_id,
provider: "email",
isSocial: false,
```

Then the rest of the elements are the same but `profileData` is required... I wonder if that's possible in TypeScript!

File: integration-test/management-api/users.spec.ts [hunk 17]
```diff
@@ -95,5 +97,112 @@ describe("users", () => {
...
+        },
+      });
+
+      expect(listUsersResponse.status).toBe(200);
+
+      const usersList = await listUsersResponse.json();
+      expect(usersList.length).toBe(1);
+      expect(usersList[0].user_id).toBe(newUser2.user_id);
+
+      // Fetch a single users
```
Comment (danielgent) [inline integration-test/management-api/users.spec.ts:177]:
```suggestion
      // Fetch a single user
```

Comment (danielgent) [reply integration-test/management-api/users.spec.ts:177]: But should it be `fetch the primary user`?

And what happens with the linked (secondary) user? Can these be fetched in auth0?  they don't appear in the `get all users` list, so we could assert for that (we are but only implicitly)

File: src/routes/management-api/users.ts [hunk 18]
```diff
@@ -51,28 +51,35 @@ export class UsersMgmtController extends Controller {
...
-        last_ip: "",
-        last_login: "",
-        identities,
+        identities: [
+          {
+            connection: user.connection,
+            provider: user.provider,
+            user_id: user.id,
+            isSocial: user.is_social,
+          },
```
Comment (danielgent) [inline src/routes/management-api/users.ts:79]: But still `TODO` is the megajoin here right to put all the other identities in? :sunglasses:

File: src/routes/management-api/users.ts [hunk 19]
```diff
@@ -96,24 +103,36 @@ export class UsersMgmtController extends Controller {
...
+      per_page: 10,
+      include_totals: false,
+      q: `linked_to:${user_id}`,
+    });
+
+    const identities = [user, ...linkedusers.users].map((u) => ({
+      connection: u.connection,
+      provider: u.provider,
+      user_id: u.id,
+      isSocial: u.is_social,
```
Comment (danielgent) [inline src/routes/management-api/users.ts:128]: and `profileData`! :smile: 

Something like these

image

**Review by danielgent — APPROVED** [review]

**Pull request merged**